HW1/P5/P5_bfs_v3.py

```diff
@@ -0,0 +1,132 @@
+
+links = sc.textFile('s3://Harvard-CS205/wikipedia/links-simple-sorted.txt')
+page_names = sc.textFile('s3://Harvard-CS205/wikipedia/titles-sorted.txt')
+#links = sc.textFile('/Users/haosutang/links-simple-sorted.txt')
+#page_names = sc.textFile('/Users/haosutang/titles-sorted.txt')
+neighbor_graph = links.map(link_string_to_KV)
+page_names = page_names.zipWithIndex().map(lambda (n, id): (id + 1, n))
+page_names = page_names.sortByKey().cache()
+
+# find Kevin Bacon
+Kevin_Bacon = page_names.filter(lambda (K, V): V == 'Kevin_Bacon').collect()
+# This should be [(node_id, 'Kevin_Bacon')]
+assert len(Kevin_Bacon) == 1
+Kevin_Bacon = Kevin_Bacon[0][0]  # extract node id
+
+# find Harvard University
+Harvard_University = page_names.filter(lambda (K, V):
+                                       V == 'Harvard_University').collect()
+# This should be [(node_id, 'Harvard_University')]
+assert len(Harvard_University) == 1
+Harvard_University = Harvard_University[0][0]  # extract node id
+
+neighbor = neighbor_graph.cache()
+
+
+
+
+
+start = Kevin_Bacon
+end = Harvard_University
+max_iteration=10
+accum = sc.accumulator(-1)
+globallevel = accum.value
+#add a level flag to the end, never visited is max_iteration, current
+#also add a origin mark, start node is 0, never visited is -1
+
+
+##################################################
+# Find the tree containing the path
+##################################################
+mapped = neighbor.map(lambda l: (l[0], (l[1], max_iteration, -1)) if l[0]!=start else (l[0], (l[1], globallevel, 0)))
+
+tree = None
+
+while globallevel<max_iteration:
+    accum.add(1)
+    globallevel = accum.value
+    print 'global level', globallevel
+
+    transition = mapped.flatMap(extractNodes)
+
+    mapped = transition.groupByKey().map(lambda l: (l[0], refreshNodeComb(list(l[1]))))
+
+    if not mapped.filter(lambda(k,v): k==end and v[1]!=max_iteration).isEmpty():# if visited
+        print '**********************************************'
+        print '******************Found!**********************'
+        print '**********************************************'
+        tree = mapped.filter(lambda (k,v): v[1]!=max_iteration and (k==end or v[1]<=globallevel))
+        break
+    mapped.persist()
+
+##################################################
+# Trace back the tree to find a path
+##################################################
+assert tree!=None
+path = [end]
+key=end
+while key!=start:
+    origin = tree.filter(lambda (k,v): k ==key).collect()[0][1][2]
+    path += [origin]
+    tree = tree.filter(lambda (k,v): k==origin or v[1]<=globallevel)
+    tree.persist()
+    key = origin
+    globallevel -= 1
+    
+
+shortest_path_id = []
+
+for i in path:
+	shortest_path_id += [page_names.filter(lambda (K, V): K==i).collect()]
+
+print '********************PATH**********************'
+print list(reversed(shortest_path_id))
+print '**********************************************'
```

Comment:
We have the same results, except your code seems slower. Also I got the general idea of your approach but your code is long and hard to follow.